custom/icds/models.py

```diff
@@ -0,0 +1,163 @@
+        return supporting_file.utility.file_exists()
+
+
+class HostedCCZ(models.Model):
+    link = models.ForeignKey(HostedCCZLink, on_delete=models.CASCADE)
+    app_id = models.CharField(max_length=255, null=False)
+    version = models.IntegerField(null=False)
+    profile_id = models.CharField(max_length=255, blank=True)
+    file_name = models.CharField(max_length=255, blank=True)
+
+    class Meta:
+        unique_together = ('link', 'app_id', 'version', 'profile_id')
+
+    @cached_property
+    def utility(self):
+        return HostedCCZUtility(self)
+
+    @cached_property
+    def domain(self):
+        return self.link.domain
+
+    def to_json(self, app_names):
+        from custom.icds.serializers import HostedCCZSerializer
+        return HostedCCZSerializer(self, context={'app_names': app_names}).data
+
+    @cached_property
+    def blob_id(self):
+        assert self.app_id
+        assert self.version
+        return "%s%s%s" % (self.app_id, self.version, self.profile_id)
+
+    @cached_property
+    def build_doc(self):
+        if self.link_id and self.app_id and self.version:
+            return get_build_doc_by_version(self.domain, self.app_id, self.version)
+
+    @cached_property
+    def build_profile(self):
+        if self.profile_id and self.build_doc:
+            return self.build_doc['build_profiles'].get(self.profile_id)
+
+    def clean(self):
+        if not self.build_doc:
+            raise ValidationError({
+                'version': _("Build not found for app {} and version {}.").format(
+                    self.app_id, self.version
+                )
+            })
+        if not self.build_doc['is_released']:
+            raise ValidationError({
+                'version': _("Version not released. Please mark it as released.")})
+        if not self.file_name:
+            self.file_name = "%s-v%s" % (self.build_doc['name'], self.version)
+        super(HostedCCZ, self).clean()
+
+    def save(self, *args, **kwargs):
+        from custom.icds.tasks.hosted_ccz import setup_ccz_file_for_hosting
+        self.full_clean()
+        super(HostedCCZ, self).save(*args, **kwargs)
+        if not self.utility.file_exists():
+            setup_ccz_file_for_hosting.delay(self.pk)
+
+    def delete_ccz(self):
+        # if no other link is using this app+version+profile, delete the file from blobdb
+        if not (HostedCCZ.objects.filter(app_id=self.app_id, version=self.version, profile_id=self.profile_id)
+                .exclude(link=self.link).exists()):
+            self.utility.remove_file_from_blobdb()
+
+    def delete(self, *args, **kwargs):
+        self.delete_ccz()
+        super(HostedCCZ, self).delete(*args, **kwargs)
+
+
+def delete_ccz_for_link(sender, instance, **kwargs):
+    for hosted_ccz in HostedCCZ.objects.filter(link=instance):
+        hosted_ccz.delete_ccz()
+
+
+pre_delete.connect(delete_ccz_for_link, sender=HostedCCZLink)
```

Comment:
I commented on this before but can't find the comment now. Is this not handled by `on_delete=CASCADE` of the foreign key?

Reply:
@snopoke 
on_delete=CASCADE does not call the delete method on the related object but just deteles it directly from DB. We need some clean up to be done when the related object is deleted.

Here is the [commit](https://github.com/dimagi/commcare-hq/pull/24129/commits/2b905b38c2199971ad033d106199cbbd4088327e) where i switched to use signal instead of overriting the delete method on CCZHostingLink after your earlier comment

Reply:
here is your comment from before https://github.com/dimagi/commcare-hq/pull/24129#discussion_r283323558